7. Typescript

ivandashk opened the pull request:
# Запуск
```
gulp
```

- Подключен TSLint
- Компиляция в строгом режиме с noImplicitAny и strictNullChecks
- Type assertions были заменены на метод cast (где возможно), которая кинет исключение о неверном типе, если пришел не тот тип, который ожидался (к сожалению, в рантайме).
Метод доступен в src/utils/ts-cast.ts

### Ответы на вопросы из задания
> Трудоёмкость перевода проекта на TypeScript. Самые сложные моменты в работе.

Самым сложным моментом считаю типизацию данных для внешних объектов (events.json), которые необходимо приводить при помощи type guard'ов и partial классов.

> Какие в процессе перевода были найдены ошибки.

Архитектурные. Строгая типизация подсказывает какие части проекта стоит вынести в отдельный модуль и какие параметры передавать в методы.

> Решили ли вы вливать данный PR, или предпочитаете работать с JavaScript? Почему?

Далее продолжу работать с TypeScript, потому что это удобно. В настоящее время наблюдаю все больше проектов на данном наречии и хочу продолжить его изучение)

Помимо типизации был произведен рефакторинг кода:
### Задание 1. Шаблонизация карточек
**Было:** скрипт card-generator.js  
**Стало:** 
- card-main.ts - получение events.json и запуск генерации
- card-generator.ts - логика генерации карточек
- card-event.ts - класс CardEvent, сущность всей карточки. Свойство data принимает тип, наследуемый от абстрактного типа CardData.
- CardDataButtons.ts - дополнение для карточки, содержащее кнопки.
- CardDataGraph.ts - дополнение для карточки, содержащее график.
- CardDataImage.ts - дополнение для карточки, содержащее картинку.
- CardDataMeasurements.ts - дополнение для карточки, содержащее данные об измерениях.
- CardDataMusic.ts - дополнение для карточки, содержащее музыкальный плеер.

### Задание 2. Сенсорный ввод 
Было: gestures-camera.js
Стало: 
- gestures-main.ts - определяет является ли устройство сенсорным и вызывает GesturesSetter.ts
- gestures-setter.ts - класс GesturesSetter, которые определяет поведение DOM-элемента, который должен реагировать на жесты
- CurrentGesture.ts - класс, содержащий тип текущего жеста
- Pointer.ts - класс, содержащий информацию о каждом указателе в жесте

### Задание 3. Мультимедиа
Было: video-page.js, active-video-module.js
Стало: 
- video-page.ts: инициализация видео
- ActiveVideo.ts: Поведение текущего активного видео
- CustomSettings.ts:  Пользовательские настройки для всех видео
- SettingsSet.ts: набор пользовательских настроек отдельного видео
- AudioData.ts: Пользовательские настройки аудиоконтекста.

ivandashk committed:
Настройка typescript под gulp + nav.ts + sidevar.ts

ivandashk committed:
Card-generator.ts

ivandashk committed:
Рефакторинг: генерация карточек

ivandashk committed:
Мультимедиа переведена на ts

ivandashk committed:
Функция cast для генерации исключений, если тип является null

ivandashk committed:
Избавление от type assertions при инициализации карточек

ivandashk committed:
Удаление @types/webaudioapi, добавление @types/hls.js

ivandashk committed:
Сенсорный ввод переведен на ts

ivandashk committed:
Форматирование с TSLint

ivandashk committed:
Удалены старые файлы

Allarepossible commented on src/blocks/card/card-generator.ts:
```diff
@@ -0,0 +1,259 @@
+    header.appendChild(close);
+};
+
+const trimTitle = (str: string, cardSize: string) => {
```
не везде указан тип выходных данных

Allarepossible commented on src/blocks/card/card-generator.ts:
```diff
@@ -0,0 +1,259 @@
+    brightness.innerHTML = "Яркость: 100%";
+};
+
+export default (events: CardEvent[]) => {
```
вот это не очень хорошая практика - лучше вынести в отдельную функцию, назвать ее осмысленно и экспортировать ее, указывая, что же она возвращает

при этом функция по хорошему должна что-нибудь возвращать и не менять входные данные

ivandashk replied:
@Allarepossible 
> при этом функция по хорошему должна что-нибудь возвращать и не менять входные данные

Я, вроде, не меняю входные данные, только на основе них мы создаем CardEvent'ы. Да, лучшим решением тут было бы, переписать все на template и вставлять в DOM в самом конце, но это больше к первому заданию, чем к этому, мне кажется)

Allarepossible replied:
понимаю, просто что увидела, чтобы в будущем поправить и писать по правильному)

Allarepossible commented on src/blocks/nav/nav.ts:
```diff
@@ -0,0 +1,19 @@
+const navItems = document.getElementsByClassName("nav__item");
+const navs = document.getElementsByClassName("nav");
+
+for (const nav of navs) {
```
лучше обернуть это в функцию или модуль, а то получается много глобальных переменных

Allarepossible replied:
и как раз добавить типизацию)

Allarepossible commented on src/blocks/video-page/video-page.ts:
```diff
@@ -0,0 +1,48 @@
+    });
+};
+
+const getAudioContext = (obj: any): obj is {
```
any - не надо использовать, или же в комментариях писать что там может быть
и кажется не хватает после is

ivandashk replied:
@Allarepossible 
Это type guard, он проверяет есть ли у объекта webkitAudioContext или AudioContext. Он должен принимать any, иначе будет ошибка, что объект не содержит двух свойств сразу.
Пример из лекции: https://github.com/shyr1punk/typescript-overview/blob/master/2_any_unknown.ts

Или можно его по другому записать?

Allarepossible replied:
оу, если пример из лекции, то тогда нормально. видимо везде разные соглашения как писать типизацию)

Allarepossible commented on tsconfig.json:
```diff
@@ -0,0 +1,59 @@
+    // "strictFunctionTypes": true,           /* Enable strict checking of function types. */
+    // "strictPropertyInitialization": true,  /* Enable strict checking of property initialization in classes. */
+    // "noImplicitThis": true,                /* Raise error on 'this' expressions with an implied 'any' type. */
+    // "alwaysStrict": true,                  /* Parse in strict mode and emit "use strict" for each source file. */
```
хитро)

ivandashk replied:
@Allarepossible ?

Allarepossible replied:
ну ты просто закомментировал большую часть правил, чтобы все проверки проходили) или вам на лекции так сказали?

ivandashk replied:
Относительно строгого режима, в примере на лекции включались только первые три правила и, вроде как, было объявлено, что этого достаточно)

Allarepossible reviewed (commented):
в половине файлов не все типизировано и не у всех функций нет типизации возвращаемых значений

ivandashk committed:
Изменены правила компиляции

ivandashk committed:
Явное указание типов возвращаемых значений

Для функций и методов, где нет оператора return возвращаемое значение void опускается

ivandashk committed:
Явное экспортирование generateCards()

ivandashk committed:
Добавлены функции-инициализаторы nav.ts и sidebar.ts

ivandashk commented:
@Allarepossible upd: 
- Исправлены замечания
- Включены все правила строгой компиляции
- Исправлены отсутствующие явные указания типов возвращаемых значений. Для функций и методов, где нет оператора return возвращаемое значение void опускается
- Для дальнейшего рефакторинга, инициализацию страниц (initSidebar(), setGestures(), generateCards()) можно вынести в отдельные файлы, но требуется подключить бандлер (webpack, browserify), чтобы импортировать hls. В данном тикете считаю эти изменения излишними.
